lib/streamlit/type_util.py

```diff
@@ -97,18 +97,30 @@
     "trigger_value",
 ]
 
+V_co = TypeVar("V_co", covariant=True)
+
+
+class DataFrameGenericAlias(Protocol[V_co]):
+    """Technically not a GenericAlias, but serves the same purpose in
+    OptionSequence below, in that it is a type which admits DataFrame,
+    but is generic. This allows OptionSequence to be a fully generic type,
+    significantly increasing its usefulness.
+
+    We can't use types.GenericAlias, as it is only available from python>=3.9,
+    and isn't easily back-ported."""
+
+    @property
+    def iloc(self) -> _iLocIndexer:
+        ...
+
+
 OptionSequence: TypeAlias = Union[
-    Sequence[Any],
-    "DataFrame",
-    "Series",
-    "Index",
-    "np.ndarray",
+    Iterable[V_co],
```

Comment:
In the cases where `OptionSequence` is used, it is always passed through `ensure_indexable`, which just requires the value to be `Iterable`, not `Sequence`. `Iterable` covers all of `DataFrame`, `Series`, `Index, `np.ndarray`, meaning `OptionSequence` could almost be just `Iterable`.

But, the special treatment that `DataFrame` gets in `ensure_indexable`, means that you don't get  the `V_co` values out, which would be the column names (bounded by `Hashable`), but rather the values of the first column. For now, that doesn't matter, as `DataFrame` is untyped and equates to `Any`, but that might change.

That is why I also include the `DataFrameGenericAlias` type, to underline the special case for DataFrames. It might be too much. Would love feedback on this.

The other option, the way I see it, is to just go with `Iterable[V_co]` and deal with the issues later if `DataFrame` ever becomes typed.

Reply:
If I remember correctly, the special-casing of dataframes was an intentional product decision, and if we wanted to reverse it at this point, it would constitute a breaking change, so the earliest it could possibly be done would be in the next major version upgrade (which doesn't have any set date at the moment).

I think defining `DataFrameGenericAlias` here is likely fine. I don't think it's ideal, but also having `DataFrame` becoming typed at some unknown point in the future and breaking things in a way that may be hard to debug sounds like a worse scenario